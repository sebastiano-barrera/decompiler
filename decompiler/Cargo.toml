[package]
name = "decompiler"
version = "0.1.0"
edition = "2021"

[features]
trace = []
file_lock = []

[dependencies]
anyhow = "1.0.89"
bimap = "0.6.3"
gimli = { version = "0.31.1", default-features = false, features = [
    "read-all",
] }
goblin = "0.8.2"
smallvec = "1.13.2"
thiserror = "2.0.9"
rayon = "1.10.0"
slotmap = "1.0.7"
enum-assoc = "1.2.4"
arrayvec = "0.7.6"

# must be exactly 0.27.0, as following versions have MSRV 1.87.0, and we're on 1.86
facet = "=0.27.0"
facet-derive = "=0.27.0"
facet-reflect = "=0.27.0"
tracing = "0.1.41"
serde = { version = "1.0.228", features = ["derive", "rc"] }
rmp-serde = "1.3.0"

[dependencies.iced-x86]
version = "1.21.0"
default-features = false
features = ["decoder", "intel", "std", "instr_info"]

[dev-dependencies]
insta = { version = "1.40.0", default-features = false }
iced-x86 = { version = "1.21.0", features = ["encoder", "code_asm"] }
include_dir = "0.7.4"

# list ourselves here, to add the 'trace' feature when we run `cargo test` by default
decompiler = { path = ".", features = ["trace"] }
test-log = { version = "0.2.18", features = ["trace"] }

[[test]]
name = "test_tool"
path = "tests/test_tool.rs"
required-features = ["trace"]
