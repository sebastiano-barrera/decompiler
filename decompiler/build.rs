use std::process::Command;

fn main() {
    const SCRIPT_GEN_CALLCONV_TESTS: &str = "scripts/gen_callconv_tests.rb";

    println!("cargo::rerun-if-changed={}", SCRIPT_GEN_CALLCONV_TESTS);
    Command::new(SCRIPT_GEN_CALLCONV_TESTS)
        .arg("--out-c")
        .arg("test-data/x86_64_callconv.c")
        .arg("--out-rs")
        .arg("tests/generated_callconv.rs")
        .status()
        .unwrap();

    // generate the test executable
    //
    // don't use the `cc` crate because we really want to use GCC and this
    // specific set of flags
    Command::new("gcc")
        // higher levels of optimizations completely hide the calling
        // conventions
        .arg("-O1")
        .arg("-gdwarf")
        .arg("-o")
        .arg("test-data/x86_64_callconv")
        .arg("test-data/x86_64_callconv.c")
        .status()
        .unwrap();
}
