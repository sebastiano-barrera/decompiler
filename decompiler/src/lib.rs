mod ast;
mod cfg;
mod elf;
#[macro_use]
mod mil;
mod ssa;
mod tests;
mod ty;
mod x86_to_mil;

// temporarily disabled
#[cfg(any())]
mod xform;
mod xform {
    use crate::ssa;
    pub fn canonical(prog: &mut ssa::Program) {
        // do nothing!
    }

    pub fn eliminate_dead_code(prog: &mut ssa::Program) {
        let mut is_read = ssa::RegMap::for_program(prog, false);

        for bid in prog.cfg().block_ids_postorder() {
            for reg in prog.block_regs(bid).rev() {
                let mut insn = prog[reg].get();

                if insn.has_side_effects() {
                    is_read[reg] = true;
                }

                if is_read[reg] {
                    for &mut input in insn.input_regs() {
                        is_read[input] = true;
                    }
                }
            }
        }

        for (reg, &is_read) in is_read.items() {
            use crate::mil;
            if !is_read {
                prog[reg].set(mil::Insn::Void);
            }
        }
    }
}

mod util;

pub mod api;
pub mod pp;

pub use api::*;
