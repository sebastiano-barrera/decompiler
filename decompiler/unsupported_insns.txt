adc r9,0
adc rdx,0
addsd xmm0,xmm0
addsd xmm0,xmm1
addss xmm0,xmm0
addss xmm0,xmm1
addss xmm1,xmm1
bsr r8,rdi
bsr rax,rdx
bsr rcx,rax
bsr rcx,rdx
bswap eax
bswap esi
bswap r10d
bswap r11d
bswap r8d
bswap r9d
bswap rax
bswap rdi
bswap rsi
bt rax,rcx
bt rcx,rax
bt rdx,2Fh
btr r8,34h
btr rax,34h
btr rdx,2Fh
bt rsi,rcx
bts qword ptr [r14+8],29h
bts qword ptr [r14+8],2Ch
bts qword ptr [r14+8],31h
bts qword ptr [r15+8],2Fh
bts qword ptr [rbx+8],28h
bts qword ptr [rbx+8],2Fh
bts qword ptr [rdi+8],28h
bts rax,30h
bts rcx,35h
cmova eax,ecx
cmova eax,edx
cmovae rax,r8
cmovae rax,rdi
cmovae rax,rdx
cmova esi,eax
cmova r12d,ebx
cmovb ebx,eax
cmovbe r12,rax
cmovbe rbx,rax
cmovbe rdx,r12
cmovbe rsi,rdx
cmovb rsi,rdi
cmove eax,[rdx+918h]
cmove edx,eax
cmove edx,ecx
cmove r12d,eax
cmove r12,r15
cmove r12,rax
cmove rax,r12
cmove rax,rcx
cmove rax,rdx
cmove rbx,r14
cmove rbx,r15
cmove rbx,rax
cmove rcx,rax
cmove rdi,rax
cmove rdx,rax
cmove rsi,rax
cmovg eax,edx
cmovg ecx,eax
cmovge eax,edx
cmovge r13d,r15d
cmovge rax,rdx
cmovg r13d,ebx
cmovg r15d,eax
cmovl eax,esi
cmovl ecx,eax
cmovl edx,eax
cmovle ecx,eax
cmovle esi,edx
cmovle esi,r13d
cmovl r15d,eax
cmovne eax,ecx
cmovne eax,edi
cmovne eax,edx
cmovne eax,r15d
cmovne ecx,eax
cmovne ecx,edx
cmovne edx,eax
cmovne edx,ecx
cmovne edx,r12d
cmovne r12d,eax
cmovne r13d,ecx
cmovne r15d,eax
cmovne rax,rbx
cmovne rdi,rbx
cmovne rdx,rax
cmovne rsi,[77A3A0h]
cmovne rsi,rax
cmovne rsi,rbx
cmovns eax,ecx
cmovns eax,edx
cmovns rax,rbx
cmovns rax,rdx
cmovns rdx,rdi
cmovs rax,rdx
comisd xmm0,xmm1
comisd xmm1,xmm0
comisd xmm1,xmm2
comiss xmm0,[682F88h]
cvtsi2sd xmm0,dword ptr [78F944h]
cvtsi2sd xmm0,qword ptr [rdi+8]
cvtsi2sd xmm0,rax
cvtsi2sd xmm0,rdi
cvtsi2sd xmm1,eax
cvtsi2sd xmm2,eax
cvtsi2ss xmm0,rax
cvtsi2ss xmm0,rdx
cvtsi2ss xmm1,r13
cvtsi2ss xmm1,rax
cvtsi2ss xmm1,rcx
cvtsi2ss xmm1,rdx
div rcx
div rdi
divsd xmm0,[rsi*8+6E32A0h]
divsd xmm1,xmm0
divsd xmm2,[69E4B0h]
divss xmm0,xmm1
fld st,tbyte ptr [rbp+10h]
fld st,tbyte ptr [rbp-30h]
fldz
fstp st(0),st
fstp tbyte ptr [rbp+10h],st
idiv dword ptr [78E634h]
lock xadd [779E90h],rax
lock xadd [779E90h],rcx
lock xadd [779E90h],rdx
movdqa xmm0,[6DD150h]
movdqa xmm0,[6DD180h]
movdqa xmm0,[6DD190h]
movdqa xmm0,[6DD1A0h]
movdqa xmm0,[6DD200h]
movdqa xmm0,[75EE60h]
movdqa xmm0,[75EE80h]
movdqa xmm0,[75EE90h]
movdqa xmm0,[75EEA0h]
movdqa xmm0,[rbp-140h]
movdqa xmm0,[rbp-160h]
movdqa xmm0,[rbp-20h]
movdqa xmm0,[rbp-40h]
movdqa xmm0,xmm1
movdqa xmm0,xmm2
movdqa xmm1,[6DD200h]
movdqa xmm1,[75EE50h]
movdqa xmm1,[rbp-40h]
movdqa xmm1,xmm0
movdqa xmm1,xmm3
movdqa xmm2,[75EE40h]
movdqa xmm2,xmm1
movdqa xmm3,xmm0
movdqu xmm0,[r10]
movdqu xmm0,[r11]
movdqu xmm0,[r12+28h]
movdqu xmm0,[r12+38h]
movdqu xmm0,[r12+60h]
movdqu xmm0,[r12+80h]
movdqu xmm0,[r14+18h]
movdqu xmm0,[r14+28h]
movdqu xmm0,[r14+38h]
movdqu xmm0,[r14+48h]
movdqu xmm0,[r14+58h]
movdqu xmm0,[r14+68h]
movdqu xmm0,[r14+78h]
movdqu xmm0,[r14+8]
movdqu xmm0,[r9]
movdqu xmm0,[rax+18h]
movdqu xmm0,[rax+8]
movdqu xmm0,[rbx]
movdqu xmm0,[rbx+14h]
movdqu xmm0,[rbx+30h]
movdqu xmm0,[rcx]
movdqu xmm0,[rdi+18h]
movdqu xmm0,[rdx]
movdqu xmm0,[rdx+18h]
movdqu xmm0,[rdx+rax*2]
movdqu xmm0,[rsi]
movdqu xmm0,[rsi+10h]
movdqu xmm0,[rsi+20h]
movdqu xmm0,[rsi+30h]
movdqu xmm1,[r12+18h]
movdqu xmm1,[r12+50h]
movdqu xmm1,[r12+70h]
movdqu xmm1,[rdx]
movdqu xmm2,[r12+8]
movdqu xmm3,[rdx+rax*2+10h]
movhlps xmm1,xmm0
movhps xmm0,[68DD80h]
movhps xmm0,[68DD90h]
movhps xmm0,[68DDA0h]
movhps xmm0,[68DDB0h]
movhps xmm0,[68DDC0h]
movhps xmm0,[68DDD0h]
movhps xmm0,[68DDE0h]
movhps xmm0,[68DDF0h]
movhps xmm0,[68DE00h]
movhps xmm0,[69E158h]
movhps xmm0,[69E168h]
movhps xmm0,[69F7C0h]
movhps xmm0,[69F7D0h]
movhps xmm0,[69F7E0h]
movhps xmm0,[r14+8]
movhps xmm0,[rax+20h]
movhps xmm0,[rax-30h]
movhps xmm0,[rbp-140h]
movhps xmm0,[rbx*8+799760h]
movhps xmm1,[69F7D0h]
movq r12,xmm0
movq r15,xmm1
movq r9,xmm0
movq rax,xmm0
movq rax,xmm1
movq rdi,xmm1
movq xmm0,[68DD78h]
movq xmm0,[68DD88h]
movq xmm0,[68DD98h]
movq xmm0,[68DDA8h]
movq xmm0,[68DDB8h]
movq xmm0,[68DDC8h]
movq xmm0,[68DDD8h]
movq xmm0,[68DDE8h]
movq xmm0,[68DDF8h]
movq xmm0,[69E150h]
movq xmm0,[69E160h]
movq xmm0,[69F7B8h]
movq xmm0,[69F7C8h]
movq xmm0,[69F7D8h]
movq xmm0,r12
movq xmm0,[r12]
movq xmm0,r13
movq xmm0,r14
movq xmm0,[r14+8]
movq xmm0,r15
movq xmm0,rax
movq xmm0,[rax+8]
movq xmm0,rbx
movq xmm0,rdx
movq xmm1,[69F7C8h]
movq xmm1,r12
movq xmm1,[r12+8]
movq xmm1,r15
movq xmm1,rax
movq xmm1,rbx
movq xmm1,[rbx+8]
movq xmm2,r13
movq xmm3,rbx
movq xmm4,r13
movsxd r12,r12d
movsxd r12,[rax+58h]
movsxd r12,[rbp+4Ch]
movsxd r12,[rbp+54h]
movsxd r13,eax
movsxd r13,ebx
movsxd r13,edx
movsxd r13,r13d
movsxd r13,r14d
movsxd r13,r15d
movsxd r13,[rbp-104h]
movsxd r14,eax
movsxd r14,ebx
movsxd r14,[rbp-104h]
movsxd r14,[rbp-74h]
movsxd r14,[rsi+0D0h]
movsxd r15,eax
movsxd r15,r12d
movsxd r15,[rbp+10h]
movsxd r8,r12d
movsxd r8,r8d
movsxd r9,r9d
movsxd rax,ecx
movsxd rax,edx
movsxd rax,esi
movsxd rax,r10d
movsxd rax,[r12+18h]
movsxd rax,[r12+50h]
movsxd rax,r12d
movsxd rax,[r13]
movsxd rax,[r13+18h]
movsxd rax,r13d
movsxd rax,[r14]
movsxd rax,r14d
movsxd rax,[r15]
movsxd rax,r15d
movsxd rax,r9d
movsxd rax,[rax*4+6832C0h]
movsxd rax,[rbp-0C0h]
movsxd rax,[rbp-44h]
movsxd rax,[rbp-78h]
movsxd rax,[rbx]
movsxd rax,[rdx*4+6832C0h]
movsxd rbx,ebx
movsxd rbx,esi
movsxd rcx,[78F948h]
movsxd rcx,ebp
movsxd rcx,ecx
movsxd rcx,esi
movsxd rcx,r10d
movsxd rdi,[78F948h]
movsxd rdi,eax
movsxd rdi,ebp
movsxd rdi,edi
movsxd rdi,esi
movsxd rdi,r12d
movsxd rdi,[r13+rbx*8+4]
movsxd rdi,[r9+0D0h]
movsxd rdi,[rbp-60h]
movsxd rdx,[78E730h]
movsxd rdx,eax
movsxd rdx,ebp
movsxd rdx,ebx
movsxd rdx,edx
movsxd rdx,r10d
movsxd rdx,r12d
movsxd rdx,r13d
movsxd rdx,[r14+0D0h]
movsxd rdx,r8d
movsxd rdx,r9d
movsxd rdx,[rbp-10Ch]
movsxd rdx,[rbx]
movsxd rdx,[rbx+40h]
movsxd rdx,[rdx+0D0h]
movsxd rsi,eax
movsxd rsi,edi
movsxd rsi,edx
movsxd rsi,esi
movsxd rsi,r13d
movsxd rsi,[r14+2F8h]
movsxd rsi,[r15+0D0h]
movsxd rsi,r8d
movsxd rsi,[rax+0D0h]
movsx eax,byte ptr [r11]
movsx eax,byte ptr [rsi]
movsx edx,byte ptr [rax]
movsx edx,byte ptr [rsi]
movsx esi,byte ptr [r15]
movsx esi,byte ptr [r15+1]
movsx esi,byte ptr [r15+2]
movsx esi,byte ptr [r15+3]
movsx esi,byte ptr [r15+4]
movsx esi,byte ptr [r15+5]
movsx esi,byte ptr [r15+6]
movsx esi,byte ptr [r15+7]
movsx esi,byte ptr [r15+8]
movsx esi,byte ptr [rbx]
movsx esi,byte ptr [rbx+1]
movsx esi,byte ptr [rbx+2]
movsx esi,byte ptr [rbx+3]
movsx esi,byte ptr [rbx+4]
movsx esi,byte ptr [rbx+5]
movsx esi,byte ptr [rbx+6]
movsx esi,byte ptr [rbx+7]
movsx esi,byte ptr [rbx+8]
movsx rax,byte ptr [rdx+32h]
movsx rax,byte ptr [rdx+33h]
movsx rdx,byte ptr [r12+rax]
movsx rdx,byte ptr [rdi]
mul qword ptr [r13*8+6E0920h]
mul qword ptr [r9*8+6E0920h]
mul qword ptr [rsp+20h]
mul r10
mul r11
mul r12
mul r13
mul r14
mul r8
mul rcx
mul rdi
mul rdx
mul rsi
mulsd xmm0,[rsi*8+6E32A0h]
mulsd xmm0,xmm1
neg eax
neg edx
neg r13
neg rax
neg rcx
neg rdi
neg rdx
neg rsi
not eax
not ecx
not esi
not r14
not rax
not rcx
not rdx
not rsi
paddq xmm0,xmm1
pcmpeqd xmm0,xmm0
pcmpeqd xmm2,xmm2
por xmm0,xmm1
por xmm1,xmm0
psrldq xmm0,8
psrldq xmm1,8
psrld xmm2,10h
psrld xmm3,8
punpckldq xmm0,xmm3
punpckldq xmm1,xmm0
punpckldq xmm1,xmm2
punpcklqdq xmm0,xmm0
punpcklqdq xmm0,xmm1
punpcklqdq xmm0,xmm2
punpcklqdq xmm0,xmm3
punpcklqdq xmm0,xmm4
rep movsq [rdi],[rsi]
rep stosq [rdi]
rol eax,1
rol eax,5
rol ebx,1
rol ebx,5
rol ecx,1
rol ecx,5
rol edi,1
rol edi,5
rol edx,1
rol edx,5
rol esi,1
rol esi,5
rol r10d,1
rol r10d,5
rol r11d,1
rol r11d,5
rol r12d,1
rol r12d,5
rol r13d,1
rol r14d,1
rol r14d,5
rol r15d,1
rol r15d,5
rol r8d,1
rol r8d,5
rol r9d,1
rol r9d,5
ror eax,2
ror ebx,2
ror ecx,2
ror edi,2
ror edx,2
ror esi,2
ror r10d,2
ror r11d,2
ror r12d,2
ror r13d,2
ror r14d,2
ror r15d,2
ror r8d,2
ror r9d,2
sar eax,10h
sar ecx,0Fh
sar ecx,8
sar edi,1
sar edi,8
sar edx,10h
sar esi,10h
sar r8,3Fh
sar rax,5
sar rcx,12h
sar rcx,3Fh
sar rdi,3Fh
sar rdx,0Eh
sar rdx,12h
sar rdx,3Fh
sar rdx,5
sar rdx,7
sar rsi,3Fh
sbb eax,0FFFFFFFFh
sbb ebx,0FFFFFFFFh
sbb edx,edx
sbb esi,0FFFFFFFFh
sbb r13,0FFFFFFFFFFFFFFFFh
sbb r13d,r13d
sbb r14,0FFFFFFFFFFFFFFFFh
sbb rax,0
sbb rax,0FFFFFFFFFFFFFFFFh
sbb rax,rax
sbb rdi,rdi
shld rax,r10,8
shld rax,rcx,34h
shufps xmm0,xmm3,88h
subss xmm2,xmm0
ucomiss xmm2,xmm0
ud2
unpcklpd xmm0,xmm1
xchg [77A2C0h],rax
xchg [77A2D8h],eax
xchg [r13],rax
xorpd xmm0,[6DD0B0h]
