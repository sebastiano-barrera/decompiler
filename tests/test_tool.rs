use decompiler::{pp::PrettyPrinter, test_tool};

use include_dir::{include_dir, Dir};
use insta::assert_snapshot;
use std::{
    path::{Path, PathBuf},
    sync::OnceLock,
};

static DATA_DIR: Dir<'_> = include_dir!("$CARGO_MANIFEST_DIR/test-data/");

macro_rules! case {
    ($case:ident, $exe:expr, $func_name:ident) => {
        #[test]
        #[allow(non_snake_case)]
        fn $case() {
            let function_name = stringify!($func_name);
            let out = $exe.process_function(function_name);
            crate::assert_snapshot!(out);
        }
    };
}

macro_rules! tests_in_binary {
    ($group:ident, $exe_path:expr; $($funcs:ident),*) => {
        mod $group {
            use crate::Exe;
            static EXE: Exe = Exe::new($exe_path);

            $(case!($funcs, EXE, $funcs);)*
        }
    };
}

struct Exe {
    once_lock: OnceLock<test_tool::Result<test_tool::Tester<'static>>>,
    path: &'static str,
}
impl Exe {
    const fn new(path: &'static str) -> Self {
        Exe {
            once_lock: OnceLock::new(),
            path,
        }
    }

    fn get_or_init(&self) -> &test_tool::Tester<'static> {
        self.once_lock
            .get_or_init(|| {
                let rel_path = Path::new(self.path);
                let raw = DATA_DIR.get_file(rel_path).unwrap().contents();
                test_tool::Tester::start(raw)
            })
            .as_ref()
            .unwrap()
    }

    fn process_function(&self, function_name: &str) -> String {
        let mut buf = Vec::new();
        let mut pp = PrettyPrinter::start(&mut buf);
        self.get_or_init()
            .process_function(function_name, &mut pp)
            .expect("function decompilation failed");

        let out = String::from_utf8(buf).unwrap();
        out
    }
}

tests_in_binary!(
    exe_composite_type, "ty/test_composite_type.so";
    list_len
);
tests_in_binary!(exe_capstone, "integration/sample_capstone"; main);
tests_in_binary!(exe_matrix, "integration/sample_matrix"; sum_matrix);
tests_in_binary!(
    redis_server, "integration/redis-server";
    ctl_lookup,
    listNext,
    listTypePush
);

tests_in_binary!(
    x86_64_callconv, "x86_64_callconv";
    func000,
    func001,
    func002,
    func003,
    func004,
    func005,
    func006,
    func007,
    func008,
    func009,
    func010,
    func011,
    func012,
    func013,
    func014,
    func015,
    func016,
    func017,
    func018,
    func019,
    func020,
    func021,
    func022,
    func023,
    func024,
    func025,
    func026,
    func027,
    func028,
    func029,
    func030,
    func031,
    func032,
    func033,
    func034,
    func035,
    func036,
    func037,
    func038,
    func039,
    func040,
    func041,
    func042,
    func043,
    func044,
    func045,
    func046,
    func047,
    func048,
    func049,
    func050,
    func051,
    func052,
    func053,
    func054,
    func055,
    func056,
    func057,
    func058,
    func059,
    func060,
    func061,
    func062,
    func063,
    func064,
    func065,
    func066,
    func067,
    func068,
    func069,
    func070,
    func071,
    func072,
    func073,
    func074,
    func075,
    func076,
    func077,
    func078,
    func079,
    func080,
    func081,
    func082,
    func083,
    func084,
    func085,
    func086,
    func087,
    func088,
    func089,
    func090,
    func091,
    func092,
    func093,
    func094,
    func095,
    func096,
    func097,
    func098,
    func099,
    func100,
    func101,
    func102,
    func103,
    func104,
    func105,
    func106,
    func107,
    func108,
    func109,
    func110,
    func111,
    func112,
    func113,
    func114,
    func115,
    func116,
    func117,
    func118,
    func119,
    func120,
    func121,
    func122,
    func123,
    func124,
    func125,
    func126,
    func127,
    func128,
    func129,
    func130,
    func131,
    func132,
    func133,
    func134,
    func135,
    func136,
    func137,
    func138,
    func139,
    func140,
    func141,
    func142,
    func143,
    func144,
    func145,
    func146,
    func147,
    func148,
    func149,
    func150,
    func151,
    func152,
    func153
);
