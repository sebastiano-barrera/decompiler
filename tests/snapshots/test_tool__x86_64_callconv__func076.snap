---
source: tests/test_tool.rs
expression: out
---
func076 0x40127d+8 (file 0x127d)
          40127d: 4c 89 c8                mov rax,r9
          401280: 48 c1 e8 20             shr rax,20h
          401284: c3                      ret

function type: "func076" func (arg0 *void <TypeID(1v1)>,
                               arg1 *void <TypeID(1v1)>,
                               arg2 small <TypeID(784v1)>,
                               arg3 small <TypeID(784v1)>) uint8_t <TypeID(788v1)>
0 warnings.


mil program = 
program  92 instrs
    0 r12 <- Ancestral(AncestralName("stack_bottom"))
    1 r2 <- Ancestral(AncestralName("CF"))
    2 r3 <- Ancestral(AncestralName("PF"))
    3 r4 <- Ancestral(AncestralName("AF"))
    4 r5 <- Ancestral(AncestralName("ZF"))
    5 r6 <- Ancestral(AncestralName("SF"))
    6 r7 <- Ancestral(AncestralName("TF"))
    7 r8 <- Ancestral(AncestralName("IF"))
    8 r9 <- Ancestral(AncestralName("DF"))
    9 r10 <- Ancestral(AncestralName("OF"))
   10 r11 <- Ancestral(AncestralName("RBP"))
   11 r12 <- Ancestral(AncestralName("RSP"))
   12 r13 <- Ancestral(AncestralName("RIP"))
   13 r14 <- Ancestral(AncestralName("RDI"))
   14 r15 <- Ancestral(AncestralName("RSI"))
   15 r16 <- Ancestral(AncestralName("RAX"))
   16 r17 <- Ancestral(AncestralName("RBX"))
   17 r18 <- Ancestral(AncestralName("RCX"))
   18 r19 <- Ancestral(AncestralName("RDX"))
   19 r20 <- Ancestral(AncestralName("R8"))
   20 r21 <- Ancestral(AncestralName("R9"))
   21 r22 <- Ancestral(AncestralName("R10"))
   22 r23 <- Ancestral(AncestralName("R11"))
   23 r24 <- Ancestral(AncestralName("R12"))
   24 r25 <- Ancestral(AncestralName("R13"))
   25 r26 <- Ancestral(AncestralName("R14"))
   26 r27 <- Ancestral(AncestralName("R15"))
   27 r28 <- Ancestral(AncestralName("ZMM0"))
   28 r29 <- Ancestral(AncestralName("ZMM1"))
   29 r30 <- Ancestral(AncestralName("ZMM2"))
   30 r31 <- Ancestral(AncestralName("ZMM3"))
   31 r32 <- Ancestral(AncestralName("ZMM4"))
   32 r33 <- Ancestral(AncestralName("ZMM5"))
   33 r34 <- Ancestral(AncestralName("ZMM6"))
   34 r35 <- Ancestral(AncestralName("ZMM7"))
   35 r36 <- Ancestral(AncestralName("ZMM8"))
   36 r37 <- Ancestral(AncestralName("ZMM9"))
   37 r38 <- Ancestral(AncestralName("ZMM10"))
   38 r39 <- Ancestral(AncestralName("ZMM11"))
   39 r40 <- Ancestral(AncestralName("ZMM12"))
   40 r41 <- Ancestral(AncestralName("ZMM13"))
   41 r42 <- Ancestral(AncestralName("ZMM14"))
   42 r43 <- Ancestral(AncestralName("ZMM15"))
   43 r44 <- Ancestral(AncestralName("memory"))
   44 r45 <- Undefined
   45 r46 <- Undefined
   46 r47 <- Undefined
   47 r48 <- Undefined
   48 r49 <- Undefined
   49 r50 <- Undefined
   50 r51 <- Undefined
   51 r52 <- Undefined
   52 r53 <- Undefined
   53 r54 <- Undefined
   54 r55 <- Undefined
   55 r56 <- Undefined
   56 r57 <- Undefined
   57 r58 <- Undefined
   58 r59 <- Undefined
   59 r60 <- Undefined
   60 r61 <- Undefined
   61 r62 <- Undefined
   62 r63 <- Undefined
   63 r64 <- Undefined
   64 r65 <- Undefined
   65 r45 <- Ancestral(AncestralName("arg0"))
   66 r14 <- Get(r45)
   67 r46 <- Ancestral(AncestralName("arg1"))
   68 r15 <- Get(r46)
   69 r47 <- Ancestral(AncestralName("arg2"))
   70 r48 <- Part { src: r47, offset: 0, size: 8 }
   71 r19 <- Get(r48)
   72 r49 <- Part { src: r47, offset: 8, size: 8 }
   73 r18 <- Get(r49)
   74 r50 <- Ancestral(AncestralName("arg3"))
   75 r51 <- Part { src: r50, offset: 0, size: 8 }
   76 r20 <- Get(r51)
   77 r52 <- Part { src: r50, offset: 8, size: 8 }
   78 r21 <- Get(r52)
0x40127d:
   79 r16 <- Get(r21)
0x401280:
   80 r46 <- Const { value: 32, size: 1 }
   81 r46 <- Widen { reg: r46, target_size: 8, sign: false }
   82 r16 <- Arith(Shr, r16, r46)
   83 r16 <- Get(r16)
   84 r6 <- SignOf(r16)
   85 r5 <- IsZero(r16)
   86 r47 <- Part { src: r16, offset: 0, size: 1 }
   87 r3 <- Parity(r47)
0x401284:
   88 r45 <- Void
   89 r45 <- Concat { lo: r45, hi: r16 }
   90 r45 <- Part { src: r45, offset: 0, size: 1 }
   91 r47 <- Ret(r45)


ssa pre-xform:
ssa program  92 instrs
.B0:    ;; .
         r80 <- Const { value: 32, size: 1 }
         r81 <- Widen { reg: r80, target_size: 8, sign: false }
         r74 <- Ancestral(AncestralName("arg3"))
         r77 <- Part { src: r74, offset: 8, size: 8 }
         r82 <- Arith(Shr, r77, r81)
         r88 <- Void
         r89 <- Concat { lo: r88, hi: r82 }
         r90 <- Part { src: r89, offset: 0, size: 1 }
         r91 <- Ret(r90)


cfg:
  entry: BlockID(0)
  BlockID(0) [0:92] -> Jmp(BlockID(1))
  BlockID(1) [92:92] -> End
  domtree:
    BlockID(0)
    |  BlockID(1)

ssa post-xform:
ssa program  92 instrs
.B0:    ;; .
         r80 <- Const { value: 32, size: 1 }
         r81 <- Widen { reg: r80, target_size: 8, sign: false }
         r74 <- Ancestral(AncestralName("arg3"))
         r77 <- Part { src: r74, offset: 8, size: 8 }
         r82 <- Arith(Shr, r77, r81)
         r90 <- Part { src: r82, offset: 0, size: 1 }
         r91 <- Ret(r90)



T0: {
  Ret ((pre:arg3[16 .. 8] >> 32 as u64)[1 .. 0]);
  end
}
